Heuristic for 'main' script selection in Bower packages

csachs opened the pull request:
Hello,

great to see the recent progress in automatic Bower / NPM - Package to WebJar generation.

When I started using automatically created Bower-package based WebJars, I however stumbled upon a problem:
The 'main' script file to be used for the RequireJS configuration is just set to the last element of the main array ( https://github.com/webjars/webjars-locator/blob/3c128a6e31d21bc1641b6fee467c6a33a919b6e5/src/main/java/org/webjars/RequireJS.java#L415 … by consecutive putting the same name).

However, in projects with multiple files in the main array, **the last file is quite often not the one which one would want to include via RequireJS**. Furthermore, the file might not even be a JavaScript file, which seems to be valid concerning the Bower documentation ( http://bower.io/docs/creating-packages/#bowerjson "The primary acting files necessary to use your package." )

In my test cases, I tried using the angular-scheme-form bower-webjar, and the resulting RequireJS used the wrong script (see https://github.com/Textalk/angular-schema-form/blob/development/bower.json … the first one has to be loaded); in another test with Angular UI Grid ( https://github.com/angular-ui/bower-ui-grid/blob/master/bower.json ) the file selected is not even a JavaScript ('./ui-grid.woff' ).

From what I see, there is currently no issue or pull-request concerning this issue, so I quickly hacked together a fix here:
Using a heuristic approach to determine the most likely main script candidate: Consider only JavaScript files, and select the one, which most closely matches the artifact name (as most often, the main script is named like the project). To allow a certain fuzzyness, the Levenshtein distances are used, instead of exact string matching. The code to compute the Levenhstein distance is taken verbatim from Apache Commons [Apache License 2.0], in order to not introduce any new dependencies.

What do you think?

I just tested the code with my two use cases, so it might be good idea to first give it a try with some more, if it is considered for inclusion.

Regards,
Christian Sachs

jamesward commented:
Thanks @csachs for putting this together!  A few things:
- I'd rather have the Apache Commons dependency than a copy&paste of the code.
- Can you add a test (or a few) that validates this?

csachs commented:
Okay, added Apache Commons as dependency and used it, added a (rather synthetic) test … but if you'd run that test against current master, it should fail.
Regards, Christian

jamesward commented:
Thanks!  I'm releasing this shortly as `0.25-1`.
